src/mqtt.c

```diff
@@ -113,7 +114,12 @@ enum MQTTErrors mqtt_init(struct mqtt_client *client,
     }
 
     /* initialize mutex */
-    MQTT_PAL_MUTEX_INIT(&client->mutex);
+
+    pthread_mutexattr_t Attr;
+    pthread_mutexattr_init(&Attr);
+    int rc = pthread_mutexattr_settype(&Attr, PTHREAD_MUTEX_RECURSIVE);
```

Comment:
This is platform specific. Platform specific code needs to be implemented in the [PAL](http://liambindle.ca/MQTT-C/group__pal.html)